### ISSUE TITLE
A primitive for suspending the event loop

### ISSUE THREAD

**Issue by oremanj** [python-trio/trio#1358]
As you'd expect for a structured concurrency system, Trio's "event loop" (`trio.run()`) always runs from start to finish -- you can't stop it partway through, do something else, and restart it again. This is definitely a good default, but there are some cases where I think a "suspend" primitive would be useful. I figured I would create an issue to figure out whether it's a good idea before doing a PR.

### The proposed primitive

Inside a Trio task, you can say `await trio.hazmat.suspend_trio_from_async(sync_fn, *args)`. This will:
* leave "Trio context" (stashing and clearing the `trio._core._run.GLOBAL_RUN_CONTEXT` thread-locals `runner` and `task`)
* leave the calling task's contextvars context
* run `sync_fn(*args)`
* reenter "Trio context" (restoring the stashed thread-locals) and the calling task's contextvars context
* return or raise (back into the task that called `suspend_trio_from_async`) whatever `sync_fn(*args)` returned or raised

In a non-task Trio context such as an `abort_fn` or IOManager implementation, you can say `trio.hazmat.suspend_trio_from_sync(sync_fn, *args)`. This works like `suspend_trio_from_async` except that it only needs to save and restore `runner`.

You can implement one in terms of the other, by taking advantage of the fact that abort_fns execute in a non-task context:
```
async def suspend_trio_from_async(sync_fn, *args):
    result = None
    def abort_fn(_):
        nonlocal result
        result = outcome.capture(suspend_trio_from_sync, sync_fn, *args)
        return trio.hazmat.Abort.SUCCEEDED
    with trio.CancelScope(shield=True) as scope:
        scope.cancel()
        await trio.hazmat.wait_task_rescheduled(abort_fn)
    return result.unwrap()

def suspend_trio_from_sync(sync_fn, *args):
    from trio._core._run import GLOBAL_RUN_CONTEXT as ctx
    assert hasattr(ctx, "runner")
    assert not hasattr(ctx, "task")
    saved_runner = ctx.runner
    del ctx.runner
    try:
        return sync_fn(*args)
    finally:
        ctx.runner = saved_runner
```

### Some ways to use it

* If you want to nest a call to `trio.run()` inside `trio.run()`, you can pass `trio.run` as the `sync_fn`. This isn't usually a good idea (since the outer `run()` will block waiting for the inner one) but for a short-lived inner task it might make sense, and be required for encapsulation reasons.

* If you're implementing a foreign event loop API without necessarily-start-to-finish `run()` semantics on top of Trio (as in `trio_asyncio.SyncTrioEventLoop`), you can use `greenlet` to switch away from the Trio run stack to a different stack, implementing that event loop's `stop()` as something that bottoms out in a suspend call with sync_fn = `some_greenlet.switch`.

* If we add pluggable IOManagers as has been suggested for #399 (qt support), and you want to support Trio+(other-opinionated-event-loop) interop, you could have an IOManager whose `handle_io()` call performs `suspend_trio_from_sync(some_greenlet.switch)` instead of sleeping. Then the other event loop would switch back to the Trio stack when Trio's nearest deadline expires or some I/O happens on an fd in which Trio is interested. (It's unclear to me whether this will work better or worse in the long run than the threaded approach suggested toward the end of #399, but it seems useful to have the necessary machinery to support both.)

* If you want to fork off a child that's not confused by appearing to still be in Trio context, you could suspend with the `sync_fn` as something like
```
def fork_child():
    if (pid := os.fork()) != 0:
        return pid
    try:
        run_child()
    finally:
        os._exit(0)
```

* It's possible to use greenlets plus this suspend primitive to make syntactically-synchronous expressions perform async operations within the same task. Details are left as an exercise to the reader.

### Possible problems

If you use this to run multiple calls to `trio.run()` in the main thread, they'll fight over the global signal state (wakeup_fd, KI handler, any other signal handlers). Can probably be resolved by saving/restoring the global signal state alongside `GLOBAL_RUN_CONTEXT.runner`.

It's weird. The things it lets you do are really weird. On the other hand, the core surface area of the change is small, and if it makes interop with other event loops easier and more robust, it might still be a net reduction in complexity for the ecosystem as a whole.

**Comment by njsmith** [comment]
Interesting idea. I think the big question is whether there are really compelling use cases that specifically need to erase the trio context.

For trio-asyncio, I think of `SyncEventLoop` as a gross internal hack for testing only. It shouldn't be a public API; its only purpose is to let us validate our implementation by running existing aio-based test suites. So if it needs some fragile special-case hacks, that doesn't bother me too much.

For #399, I think the best idea so far is to provide a version of `trio.run` that unrolls itself into a chain of callbacks, that schedule themselves on the foreign event loop. Two things about this that seem relevant here:

- if `trio.run` knows how to unroll itself into a chain of callbacks, then if you squint that actually has a lot of overlap with what you're suggesting here. The details are different, but fundamentally both approaches are ways of pausing and resuming `trio.run`.

- for event loop cohabitation, you probably *want* to preserve trio context even when trio code isn't running, so e.g. your qt callback can do `trio_event.set()`.

For fork... oof, that's messy. Ideally, I think you want the child to not just drop the trio thread locals, but actually close the `IOManager`, and maybe even drop other tasks so they can be GCed... or something? I'm really not sure what the right semantics are. But it's definitely not obvious to me that the proposal here would be sufficient for that that case.

So for me it's not clear whether this really has good use cases or not.

**Comment by smurfix** [comment]
> For trio-asyncio, I think of SyncEventLoop as a gross internal hack for testing only.
Yep. Definitely.

> unroll into a chain of callbacks
That's one option. Another is to morph _io_epoll into _io_qt and _io_gtk. (That's one for the "how hard could it possibly be" drawer. I found `trio-asyncio` in there too ;-) )

**Comment by njsmith** [comment]
> That's one option. Another is to morph _io_epoll into _io_qt and _io_gtk. (That's one for the "how hard could it possibly be" drawer. I found trio-asyncio in there too ;-) )

That's discussed in #399 too, and it's the traditional approach pioneered by Twisted. But I'd really like to avoid it if we can, because it means reimplementing all our I/O APIs repeatedly for every supported loop, putting knowledge of UI libraries directly in the Trio core, and being restricted to the minimal subset of I/O features that are supported by every loop.

**Comment by oremanj** [comment]
> For #399, I think the best idea so far is to provide a version of trio.run that unrolls itself into a chain of callbacks, that schedule themselves on the foreign event loop.

Agreed in principle. The relevant comment https://github.com/python-trio/trio/issues/399#issuecomment-406200369 mentions a few places where the "unrolled" event loop might force us into unfortunate choices between performance regressions and duplicated code, which was also my concern. But I guess if we drop the need for saving and restoring the thread-local state, you can experiment with the greenlet approach *today*, without any changes to the Trio core at all. (For a fully robust integration I think you probably need some knobs for things like "current effective deadline changed while Trio event loop wasn't running", but they're not the knobs being discussed here.)

Possibly thorny exception: greenlets are implemented (on CPython at least) by literally copying the "switched from" portion of the C stack to a heap-allocated buffer, and overwriting it with the stack of the "switched to" greenlet. This works fine with pure Python code, because all Python objects are heap-allocated. Once you add an extension module into the mix, soundness requires proving that the extension module doesn't stack-allocate any storage that's reachable from a Python object. It probably doesn't, because this would be dicey regardless, since the Python object lifetime doesn't need to obey stack discipline. But it's easy to imagine cases that would work in all practical uses currently (even though they're technically unsound) but then completely blow up when you start using greenlet with them.

> for event loop cohabitation, you probably want to preserve trio context even when trio code isn't running, so e.g. your qt callback can do trio_event.set().

Yeah, if there's only one Trio event loop in your thread, then you want to keep its context around so you can do that. You only need the save/restore if there are multiple Trio event loops in your thread. I got here via trio-asyncio, which currently has a failing test that creates multiple event loops in one thread, and when using SyncTrioEventLoop each asyncio event loop gets its own Trio event loop. But this is, like, an edge case of an edge case.

> For fork... oof, that's messy. Ideally, I think you want the child to not just drop the trio thread locals, but actually close the IOManager, and maybe even drop other tasks so they can be GCed... or something? I'm really not sure what the right semantics are. But it's definitely not obvious to me that the proposal here would be sufficient for that that case.

Good point -- fork() almost certainly wants its own thing now that I think about it.

> That's discussed in #399 too, and it's the traditional approach pioneered by Twisted. But I'd really like to avoid it if we can, because it means reimplementing all our I/O APIs repeatedly for every supported loop, putting knowledge of UI libraries directly in the Trio core, and being restricted to the minimal subset of I/O features that are supported by every loop.

On Unix-like systems where epoll and kqueue FDs themselves can be polled for readability, you can run the epoll or kqueue IOManager on top of a foreign event loop that provides a generic FD readability callback; you don't need a second thread. On Windows, #52 is done so the threaded approach mentioned in #399 sounds to me like it would work fine.
